demo-orgs: Add UI for converting to permanent organization.

laurynmm opened the pull request:
Revision of the 3rd commit in https://github.com/zulip/zulip/pull/19741.

Adds warning banner to organization settings overlay/tabs for demo organizations, which is also a button for opening the modal to convert the demo organization into a permanent organization.

---

**Notes**:
- Tim's comment in the above PR about this commit:

> The last commit can be its own independent PR; it looks like it's not too distant from the final shape that we want, but I think probably the first N rounds of iteration for it will be with @alya to get the language/flow right.

- Currently, the only updates I've made to this work is cleaning up "org" to be "organization" in settings and templates, and a small update to some jQuery formatting.

- Because we've moved the creation of demo organizations without an email to a separate PR (see #22666), the extra field on the form for adding an email address doesn't automatically appear when testing in the dev environment. I've included a screenshot and GIF of that version of the modal below, as well as a screenshot of the modal for a demo organization where the owner's email has already been set.

- Some things that need to be addressed in addition to language/flow:
  - The email update is not processed with the submit button. Though this likely will be an extension of what we do in the PR noted in the previous bullet point.
  - The dev log-in drop down for realms will need to be adjusted for the conversion of demo organizations.

--- 

**Screenshots and screen captures:**

<details>
<summary>Warning banner in organization settings modal</summary>

![Screenshot from 2022-08-16 16-22-17](https://user-images.githubusercontent.com/63245456/184912981-bc20232d-a7ce-4671-b197-9b9e3419b6b9.png)
</details>

<details>
<summary>Convert to permanent realm modal without email field</summary>

![Screenshot from 2022-08-16 16-21-28](https://user-images.githubusercontent.com/63245456/184913016-ed9168d2-2773-4001-b8d6-4200b0f8d266.png)
</details>

<details>
<summary>Convert to permanent realm modal with email field</summary>

![Screenshot from 2022-08-16 16-20-58](https://user-images.githubusercontent.com/63245456/184913072-9979147a-b83e-46df-9cb8-947db77174bc.png)
</details>

<details>
<summary>GIF of permanent realm modal with email field</summary>

![Peek 2022-08-16 16-18](https://user-images.githubusercontent.com/63245456/184913125-58fbea66-04d5-4f95-a00b-f35a851e9d2a.gif)
</details>

---

**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.

alya commented on static/js/admin.js:
```diff
@@ -86,6 +92,55 @@ function get_realm_level_notification_settings(options) {
+            html_heading: $t({defaultMessage: "Make organization permanent"}),
+            html_body,
+            on_click: submit_subdomain,
+            html_submit_button: $t({defaultMessage: "Make organization permanent"}),
```
Hm, that action button label seems a bit long/complicated. I think either a simple "Confirm" might be OK, or else "Convert". What do you guys think?

laurynmm replied:
I like "Confirm" (see screenshot below) more than "Convert". 

Looking at this, do we want that submit button to be disabled until they update the URL? I assume that no one would intentionally choose their randomly generated demo organization name. Currently, there's an error returned if you submit without changing the URL (see 2nd screenshot below).

<details>
<summary>Confirm button text screenshot</summary>

![Screenshot from 2022-08-23 16-02-27](https://user-images.githubusercontent.com/63245456/186179270-c4c8eb7a-fc1b-48ea-8bb1-4e3065e86d06.png)
</details>

<details>
<summary>Error message when URL isn't changed</summary>

![Screenshot from 2022-08-23 16-04-21](https://user-images.githubusercontent.com/63245456/186179519-70d29462-4d09-444c-9723-754f1b7285a4.png)
</details>

laurynmm replied:
With the current updates, the submit button text is "Convert".

alya replied:
With the proposal below to leave the input field blank to start with, I don't think checking whether the URL matches the current one will really come up.

Disabling the "Confirm" button until you actually enter something would be nice; we do that in a bunch of places.

alya commented on static/templates/settings/demo_organization_warning.hbs:
```diff
@@ -0,0 +1,12 @@
+    {{#if is_owner }}
+        {{#tr}}
+        Convert to make permanent.
+        {{/tr}}
```
Maybe "Convert to make it permanent." would be more clear?

laurynmm replied:
That's definitely better. Another option I thought of is "Make it a permanent organization". See two screenshots below.

<details>
<summary>Alya's updated text</summary>

![Screenshot from 2022-08-23 16-10-45](https://user-images.githubusercontent.com/63245456/186180987-be5a7a7d-0718-4516-802e-3292e44f89ca.png)
</details>

<details>
<summary>Lauryn's alternative text</summary>

![Screenshot from 2022-08-23 16-10-13](https://user-images.githubusercontent.com/63245456/186181112-f9413001-93e4-49b2-9779-c975ab1d70a7.png)
</details>

alya replied:
Hm, I think I prefer keeping "convert" as the action verb here (and elsewhere). @timabbott what do you think?

laurynmm replied:
With the current updates, the text is "Convert to make it permanent".

laurynmm commented on static/templates/settings/convert_demo_organization_form.hbs:
```diff
@@ -0,0 +1,33 @@
+              value="{{ realm_subdomain }}"
+              required />
+            <label for="string_id" class="domain_label">.{{ realm_domain }}</label>
+            <div id="help-text">The URL users will use to access the new organization.</div>
```
I occasionally read URL as an adjective describing 'users' here, which doesn't make sense. I feel like adding 'that' would just help clarify this sentence ...

> The URL that users will use to access the new organization.

alya replied:
I found this phrase odd too when I saw it, but it's the same as the wording we use on the org creation page, and we should keep the two places consistent unless there's good reason not to.

The "users will use" phrasing bothers me, so if we are going to change it, we should probably brainstorm in #design to try to get to wording we actually like.

laurynmm replied:
This was removed from the new organization form in commit 899f535b39cc3a91818e2cb658b0f9ee5baf5772.

laurynmm commented:
Noting here that currently this form and modal is not limited to the demo organization owner. So that's something to generally fix. 

But a question I have is when/if a non-owner clicks on the banner/button, what do we want to happen?

alya commented:
Oh, hm, good question. I thought it wouldn't be a link at all for non-owners, but maybe it would be better to make it go to a help page for demo orgs?

laurynmm commented:
**Updates**:
- Updated the design of the warning banner in the **Organization settings** tab to more closely match the compose box alert banner red design (which can be triggered by composing a message that exceeds the character limit). The previous design/colors weren't very good in dark theme. *See screenshots below.*
- Updated the design of the convert demo organization modal so that it is more like the other modals in the web app. The previous design was trying to mimic the design of the registration forms for creating a new organization. *See screenshots below.*
- The convert demo organization modal will only open on clicking the banner if the user is a demo organization owner.
  - Other users will see a modified version of the banner with a link to the help center article on demo organizations.
  - I'm wondering if we should only show the banner for demo organization owners and administrators. I don't think it makes much sense for other users and clutters many of the **Organization settings** tabs since they also have a banner warning that only organization admins as update the settings.
- The input field and convert button in the modal will be deactivated if the owner has not configured an email.
  - There's a warning banner in the modal in this case, currently with the same text/help link that appears in other areas of the UI that are disabled before setting an email. See screenshots below
  - I didn't change the text so as not to introduce another string to be translated, but we can definitely do so if we'd like this banner to have more specific text.
- Once the owner sets an email, the convert demo organization modal can be updated/submitted.
  - The subdomain *must* be changed. Submitting without updating it returns an error from the server (see screenshot below). Since this is the case, do we want to include the current subdomain value? Or have it blank with a placeholder? Or do we want text letting the owner know they need to change that field?
  - When successful, the owner (and all other users) will be logged out of the demo organization subdomain. They will need to log in again on the new subdomain. Do we want to indicate this will happen on the modal somewhere?

**Other Notes/Questions**:
- As I was QAing these changes, it's pretty obvious that we need to address the fact that the owner never agrees to the TOS. This is best addressed in a separate PR, but noting here so that it's on a checklist of sorts.
- Also noted when looking at the dev environment emails generated, the owner needs to get the onboarding emails. Again probably best addressed as a separate PR.
- We never resolved either of these previous comments on the modal text above: https://github.com/zulip/zulip/pull/22729#discussion_r951960128, https://github.com/zulip/zulip/pull/22729#discussion_r951960660. Will note the current state of the text in these updates in those comments/conversations.
- The production suite CI failures are likely not due to these changes.

**Updated screenshots**:

<details>
<summary>Warning banner/button in org settings - for owner</summary>

![Screenshot from 2023-09-14 16-07-34](https://github.com/zulip/zulip/assets/63245456/d86b309b-d534-4844-ae63-c4618838f693)
![Screenshot from 2023-09-14 16-07-22](https://github.com/zulip/zulip/assets/63245456/32dac5bf-5f9e-465f-adf6-f8fa68b7bff6)
</details>

<details>
<summary>Warning banner in org settings - for non-owner and non-admin</summary>

![Screenshot from 2023-09-14 16-24-58](https://github.com/zulip/zulip/assets/63245456/4e43d728-45d7-4c73-be88-0dc5c42fda69)
![Screenshot from 2023-09-14 16-24-46](https://github.com/zulip/zulip/assets/63245456/fb49444a-a00f-48fb-abeb-67a567fb40ad)
</details>

<details>
<summary>Convert demo organization modal - before setting email</summary>

![Screenshot from 2023-09-14 16-08-25](https://github.com/zulip/zulip/assets/63245456/f41376c8-c5e2-4dbf-9ed4-488a8bb69f4b)
![Screenshot from 2023-09-14 16-08-02](https://github.com/zulip/zulip/assets/63245456/12f8fceb-4ce2-4ad6-85d8-19d35ac1f605)
</details>

<details>
<summary>Convert demo organization modal - after setting email</summary>

![Screenshot from 2023-09-14 16-22-58](https://github.com/zulip/zulip/assets/63245456/2f91c678-b0c9-46b5-a20e-855058ad7c48)
![Screenshot from 2023-09-14 16-22-31](https://github.com/zulip/zulip/assets/63245456/7f4b04b4-98aa-4c06-aeaf-d77f766af0fe)
</details>

<details>
<summary>Convert demo organization modal - error from server</summary>

![Screenshot from 2023-09-14 16-23-43](https://github.com/zulip/zulip/assets/63245456/e75dce3b-55a5-47cd-a3b0-78d264b630e5)
![Screenshot from 2023-09-14 16-23-30](https://github.com/zulip/zulip/assets/63245456/5f8d5d2c-01b6-45e4-a61f-8129a271757d)
</details>

alya commented:
> I'm wondering if we should only show the banner for demo organization owners and administrators. 

Hm, yeah, that makes sense to me. Non-admins will still have the initial blue banner at the top, and it shouldn't be relevant for what non-admins do in org settings (since they're just looking anyway).

alya commented:
Playing with the PR, we need hover behavior on the "Convert to make it permanent" banner, similar to what we have on the "Upgrade to access" banner for Zulip Cloud Standard. Otherwise, it's not clear that the banner itself is a button.

alya commented:
> Since this is the case, do we want to include the current subdomain value? Or have it blank with a placeholder?

Yes, I think we want it blank with the same placeholder ("acme") we use in the new org form.

alya commented:
> When successful, the owner (and all other users) will be logged out of the demo organization subdomain. They will need to log in again on the new subdomain. Do we want to indicate this will happen on the modal somewhere?

We should add detailed documentation to the help center and link to it... But probably having a brief note directly in the modal will help too.

Will users be able to log in with the same email & password, just at the new URL?

timabbott commented:
> As I was QAing these changes, it's pretty obvious that we need to address the fact that the owner never agrees to the ToS.

The owner will need to agree to the ToS when creating the demo organization; since that's the point where they become a user of Zulip (the fact it's a special demo organization isn't' relevant). So I think that's out of scope for this PR.

timabbott commented:
> The alternative would be making just the "Convert..." part a link. I'm debating if that would be more clear.

I prefer the option of just making the "Convert" phrase a link.

laurynmm commented:
> Will users be able to log in with the same email & password, just at the new URL?

Yes, same email and password. Just at the new URL. What we're doing via converting the demo organization is basically the same as [changing the organization URL](https://zulip.com/help/change-organization-url) with the additional bit of removing the realm's deletion date.

---

**Updates**:
- The warning banner in the **Organization settings** tabs is only shown for admin users.
  - Organization owners will have the "Convert it to make it permanent" link text that opens the modal.
  - Organization admins will have the "More information" link text that opens the help center article on demo organizations in a separate tab.
  - The cursor on hover only changes for the link text.
- The modal is no longer populated with the current demo organization subdomain. Instead, the 'acme' placeholder is shown.
- The modal submit button is deactivated until the user inputs text.
- Once the user sets an email, the warning in the modal is updated to have information about users needing to log in again after converting to a new permanent URL. *See screentshot below*.
- The modal header has a link to the section in the help center article on converting a demo organization, and that help center article is updated a little bit, but could still likely be improved/expanded.

**Updated screenshots**:

<details>
<summary>Convert demo organization modal - after email is set</summary>

![Screenshot from 2023-09-18 20-13-10](https://github.com/zulip/zulip/assets/63245456/d758ba50-4e3e-4344-b426-67667367bbab)
![Screenshot from 2023-09-18 20-11-40](https://github.com/zulip/zulip/assets/63245456/6a95c626-7872-460f-bcb2-792bc03ef6c3)
</details>
<details>
<summary>Convert demo organization modal - server error for bad subdomain format</summary>

![Screenshot from 2023-09-18 20-12-38](https://github.com/zulip/zulip/assets/63245456/072b8727-a9f0-49b4-9396-b49f4dc2f9e2)
![Screenshot from 2023-09-18 20-12-04](https://github.com/zulip/zulip/assets/63245456/1b4aac35-b245-434f-8e84-69e89fee73ef)
</details>

alya commented:
Maybe something like this for the warning, to keep it short?

> All users will need to log in again at your new organization URL.

alya commented:
I think we should add a "Convert to make it permanent." link everywhere that we warn the user about the org being deleted after 30 days. I see that in the banner at the top and in the Welcome Bot message; not sure if there's anything else.

alya commented:
I'm wondering if in addition to the "more information" link, we should hint to non-owners that it's possible to make the org permanent. Not sure how important that is. @timabbott do you have thoughts on this?

laurynmm commented:
**Updates**:
- Shortened the alert in the convert demo organization modal for @alya's suggestion. See screenshot below.
- Added the owner only warning to the help center article section on converting demo organizations. This page will still need some follow-up work after this PR. See screenshot below.
- Marking PR as not a draft, but these changes may not be ready for merging. See note below about opening the conversion modal.

---

**Updated screenshots**:

<details>
<summary>Updated modal for shorter warning text - owner's email is set - light theme</summary>

![Screenshot from 2023-09-21 13-45-45](https://github.com/zulip/zulip/assets/63245456/50e7c5bb-1e54-4694-80ba-c117827eaeac)
</details>
<details>
<summary>Help center - section on converting demo organizations</summary>

![Screenshot from 2023-09-21 14-41-46](https://github.com/zulip/zulip/assets/63245456/c197c2f4-7ffb-42d0-95cb-4b2b80c2d29c)

</details>

---

**Open convert demo organization modal from navbar banner**:

> I think that for owners, we should add a "Convert to make it permanent." link everywhere that we warn the user about the org being deleted after 30 days. I see that in the banner at the top and in the Welcome Bot message; not sure if there's anything else.

To open the convert demo organization modal from outside the organization settings overlay, I think I'd need to rewrite the implementation to be more link the invite users modal, which can be opened from various places (gear menu, organization settings, right sidebar). Currently, the click handler and modal are all in `web/src/admin.js`, and not a separate `demo_organization.js` module. So, if we decide to do that, then we shouldn't merge these changes until I do that update.

And I don't know if we'd be able to open a modal from a zulip message. As far as I know, the parts of the UI that we can link to in messsages are URLs with hashes, like the various overlays. We could update the message to have a sentence/link to the convert demo organization section of the help center article, which would have the instructions via the organization settings overlay?

Another thought is when the demo organization owner sets an email, we could maybe send another message via the Welcome bot with some tips for updating the organization name/type/description before inviting users (as this impacts the onboarding experience) and more explicitly explaining how to convert the demo organization to a permanent organization (not a button /  link to the modal perse, but maybe a link to the organization settings page)?

@timabbott mentioning you to get your thoughts on the above.

---

**Blue box outline**:

> Not a blocker for merging this PR, but I noticed that if I go into the "convert" modal and click "Cancel", I see an extraneous focus box on "Convert to make it permanent" when the modal closes.
> 
> <img alt="Screen Shot 2023-09-20 at 11 53 30 AM" width="743" src="https://user-images.githubusercontent.com/2090066/269389001-64db2054-66bf-47ba-b3f8-ec27305d3702.png">

This is the outline indicating the link/button element has focus but isn't active, and I don't think we'd want to remove that for accessibility of non-mouse users. If you tab through the organization settings overlays, you'll see this outline move around. Improving this for the org/user settings overlays would be a potentially good issue/task, but probably out of the scope of this PR and demo organization work.

*And another side note is that when I tab through the settings overlay, then it will start tabbing through the left sidebar in the background and the tooltips pop up over the settings overlay.*

---

**Admin users warning banner**:

> I'm wondering if in addition to the "more information" link, we should hint to non-owners that it's possible to make the org permanent. Not sure how important that is. @timabbott do you have thoughts on this?

Would the thought be maybe to have a popver on hover with 'Only owners can convert demo organizations.'? Similar to the popover hover in the organization settings permissions tab?

Note that non-admin users will never see this banner at all.

alya commented:
> We could update the message to have a sentence/link to the convert demo organization section of the help center article, which would have the instructions via the organization settings overlay?

That sounds good! It's not worth doing something complicated here.

alya commented:
> And another side note is that when I tab through the settings overlay, then it will start tabbing through the left sidebar in the background and the tooltips pop up over the settings overlay.

Mmm, sounds worth an issue / #issues thread. :)

alya commented:
> Would the thought be maybe to have a popver on hover with 'Only owners can convert demo organizations.'? Similar to the popover hover in the organization settings permissions tab?

I think just changing the wording a bit. How about making it "...deleted in 30 days, unless it's converted into a permanent organization.[More information]()"?

timabbott commented:
> How about making it "...deleted in 30 days, unless it's converted into a permanent organization.More information"?

I like that except the "more information". Maybe better would be just for "converted into a permanent organization" to be a link to the /help/ page for users without permission, and a link to the UI for doing it for the owner?

laurynmm commented:
**Updates**:
- Added ability to open conversion modal via the keyboard (`Enter`) once it's focused.
- Adjusted the styling of the focus outline when clicked/tabbed to.
- Removed the icon at the start of the warning banner as it seems we're moving away from those in general and as the updated text is longer.
- Updated the text of the warning banner to be the same for both owners and admins (though clicking on the link text opens the modal for owners and goes to the help center article for admins). See updated screenshots below.
---

**Screenshots**:

<details>
<summary>Dark theme</summary>

![Screenshot from 2023-09-26 17-56-16](https://github.com/zulip/zulip/assets/63245456/03331531-6125-40af-9ac9-edf081f238b9)
![Screenshot from 2023-09-26 17-56-08](https://github.com/zulip/zulip/assets/63245456/79b1d318-2246-4042-9151-f1a7c6fef132)
</details>
<details>
<summary>Light theme</summary>

![Screenshot from 2023-09-26 17-56-51](https://github.com/zulip/zulip/assets/63245456/15d4313a-68e0-44be-a302-fee750b234ad)
![Screenshot from 2023-09-26 17-55-52](https://github.com/zulip/zulip/assets/63245456/4eb6d6b5-d189-4e0f-ba4f-15f01c31f023)
</details>
<details>
<summary>Dev tools simulating web app on mobile - demo organization warning banner</summary>

![Screenshot from 2023-09-26 18-04-03](https://github.com/zulip/zulip/assets/63245456/7e0fefdc-baad-413c-8483-5a211ed4a005)

</details>

---

**Remaining questions/tasks**:
- I feel like we should update the Welcome Bot initial message text in a separate PR.
- I didn't get clear feedback on whether we also wanted owners to be able to access the conversion modal from the navbar alert banner. Currently, it's still only accessible from the organization settings modal tabs.

alya commented:
Let's add ", unless it's [converted into a permanent organization](/help/demo-organizations#convert-a-demo-organization-to-a-permanent-organization)" to the banner at the top. I think it's OK to just have the help center link for everyone.

alya commented:
For organization type, let's:

1. Set it to "Unspecified" for new demo orgs.
2. Add an "Organization type" field above "Organization URL" in the field in the conversion modal (matching the order in the new org form).

The "Organization type" field should be shown regardless of whether it's been configured (blank to start with if the type is unspecified), and the "Convert" button should be disabled unless a type has been set.

Does that sound like a plan?

laurynmm commented:
**Updates**:
- Demo organizations in the dev environment are now created with the organization type as "Unspecified".
- The navbar alert for demo organizations now has the same ", unless it's converted into a permanent organization." text at the end with a link to the help center article section on converting a demo organization.
- The convert to a permanent organization modal has an additional input dropdown for organization type. The type must be changed from "Unspecified" for the submit button to be enabled.

<details>
<summary>Navbar alert</summary>

![Screenshot from 2023-09-28 18-08-23](https://github.com/zulip/zulip/assets/63245456/15762f69-f55a-4cd2-b162-f7a0ee9cedb6)
</details>
<details>
<summary>Updated modal for converting demo organization</summary>

![Screenshot from 2023-09-28 18-09-09](https://github.com/zulip/zulip/assets/63245456/4ca72db5-746d-43de-9d1c-23bd600d3d28)
![Screenshot from 2023-09-28 18-09-04](https://github.com/zulip/zulip/assets/63245456/0ac92876-13d1-474b-8db1-4b3edf5ebafd)
![Screenshot from 2023-09-28 18-08-56](https://github.com/zulip/zulip/assets/63245456/be236459-a296-4423-b8dc-d208ecf25775)
</details>

timabbott commented:
Can we drop the "Hide notice" text? It seems like the "X" does that just fine.

alya commented:
Other than Tim's comment above, the screenshots look good to me!

laurynmm commented:
**Updates**:
- Removed 'Hide notice' text/button from navbar alert banner for demo organizations.

<details>
<summary>Updated navbar banner</summary>

![Screenshot from 2023-09-28 20-07-12](https://github.com/zulip/zulip/assets/63245456/9fdd51c1-190d-476c-9fba-993f6c974928)

</details>

**Notes for code review**:
- I built these changes up off of the 3rd commit of #19741 (*which we can probably close at this point?*).
- I feel like there's a bunch of demo organization code is added to `web/src/admin.js` in these changes and wonder if it would be better to have them in a separate `web/src/demo_organizations.js` module?

timabbott commented:
This lgtm but I like your idea of putting this into a `web/src/demo_organizations.js` or maybe a `web/src/demo_organization_ui.js` file at your discretion.

laurynmm commented:
Moved the functions added for rendering the demo organization warning and convert to permanent organization modal to `web/src/demo_organizations_ui.js`. And added that new file to the exempt files for node coverage.

Otherwise, all the other code changes are the same as the previous version.

Did a quick QA review with creating a demo organization in the dev environment, adding an owner email, inviting an admin and a regular member to check the banners/links, and converting the organization to a permanent organization.
